File: rules/core/testdata/vulnerable/sql_injection.ts
import express, { Request, Response } from "express";
import Knex from "knex";

const app = express();
const knex = Knex({ client: "pg", connection: "postgresql://localhost/mydb" });

app.get("/users/search", async (req: Request, res: Response) => {
  const name = req.query.name as string;

  // VULNERABLE: string interpolation in knex.raw()
  const users = await knex.raw(`SELECT * FROM users WHERE name = '${name}'`);
  res.json(users.rows);
});

app.get("/users/:id", async (req: Request, res: Response) => {
  const id = req.params.id;

  // VULNERABLE: concatenation in raw query
  const user = await knex.raw("SELECT * FROM users WHERE id = " + id);
  res.json(user.rows[0]);
});

File: rules/core/testdata/vulnerable/sql_injection.py
from flask import Flask, request, jsonify
from sqlalchemy import create_engine, text

app = Flask(__name__)
engine = create_engine("postgresql://localhost/mydb")


@app.route("/users/search")
def search_users():
    name = request.args.get("name")

    # VULNERABLE: string interpolation in SQL query
    query = f"SELECT id, email FROM users WHERE name = '{name}'"
    with engine.connect() as conn:
        result = conn.execute(text(query))
        rows = [dict(row._mapping) for row in result]
    return jsonify(rows)


@app.route("/users/<int:user_id>/notes")
def get_notes(user_id):
    search = request.args.get("q", "")

    # VULNERABLE: f-string in text()
    with engine.connect() as conn:
        result = conn.execute(
            text(f"SELECT * FROM notes WHERE user_id = {user_id} AND body LIKE '%{search}%'")
        )
        rows = [dict(row._mapping) for row in result]
    return jsonify(rows)
